### DISCUSSION TITLE
State doesn't seem to be persisted using Nuxt 2 and `secure-ls` package

### DISCUSSION THREAD

**Discussion by theolavaux** [iendeavor/pinia-plugin-persistedstate-2#17] [Q&A] [answered]
**Describe the bug**
I'm using Nuxt 2 and I would like to persist my state using the `secure-ls` package.
If I put log in `getItem`, `setItem` or `removeItem`, none seems to be called.
And my state is not persisted after a page refresh for instance.

**To Reproduce**
[Reproduction link](https://codesandbox.io/s/stoic-breeze-kzpes1)

1. Create a Nuxt 2 project.
2. Implement basic Pinia state.
3. Install `pinia-plugin-persistedstate2`.
4. Create a file named `pinia.client.ts` with the following content.

```ts
import { Plugin } from '@nuxt/types';
import { createPersistedStatePlugin } from 'pinia-plugin-persistedstate-2';
import SecureLS from 'secure-ls';

const piniaPersistedStatePlugin = () => {
  const ls = new SecureLS({ isCompression: false });

  createPersistedStatePlugin({
    storage: {
      getItem: (key) => ls.get(key),
      setItem: (key, value) => ls.set(key, value),
      removeItem: (key) => ls.remove(key),
    },
  });
};

const piniaPlugin: Plugin = ({ $pinia }) => {
  $pinia.use(piniaPersistedStatePlugin);
};

export default piniaPlugin;
```

**Expected behavior**
State is persisted

**Comment by soc221b** [comment, marked as answer]
The `createPersistedStatePlugin` function will return an install function, we need to register it:

```diff
- const piniaPersistedStatePlugin = () => {
+ const piniaPersistedStatePlugin = (context) => {
  const ls = new SecureLS({ isCompression: false });

- createPersistedStatePlugin(/* ... */);
+ const install = createPersistedStatePlugin(/* ... */);
+ install(context);
};
```

**Reply by soc221b** [reply]
I've updated the usage section:

https://github.com/iendeavor/pinia-plugin-persistedstate-2/commit/6959cb6f6d34e64c94ee86175584e982504df793

Hope this helps you. :smile:

**Comment by theolavaux** [comment]
Awesome! It works now, only one problem left: if I clear the `localStorage`, the first time the state is fetched, I get this error 🤔 

<img width="783" alt="Screenshot 2022-03-17 at 18 15 54" src="https://user-images.githubusercontent.com/25274586/158857069-3777e447-7898-4f45-895a-8b7a1461e7f3.png">

**Reply by theolavaux** [reply]
[Here](https://gist.github.com/theolavaux/a78b22ecbd666b8d41e9d9815dae6f84) is my code

**Comment by theolavaux** [comment]
I found the culprit! I went into the debugger and in my case here the `value` equals an empty string 😢 
That's why I get an exception

https://github.com/iendeavor/pinia-plugin-persistedstate-2/blob/3b9e20421c51c1dd0aa648baaba6587e67010ac3/src/plugin.ts#L147

**Reply by soc221b** [reply]
In general, if the key does not exist, `null` or `undefined` will be returned.

- [Storage API](https://developer.mozilla.org/en-US/docs/Web/API/Storage/getItem#return_value)
- [React-native-async-storage](https://react-native-async-storage.github.io/async-storage/docs/api#getitem)
- [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/get)

For `secure-ls`, we have a workaround:

```ts
  const install = createPersistedStatePlugin({
    deserialize: (value) => {
      if (value === "") {
        return null;
      } else {
        return JSON.parse(value);
      }
    },
    // ...
```

https://codesandbox.io/s/romantic-sun-fhwi4d?file=/plugins/pinia.client.js:260-401

**Reply by theolavaux** [reply]
Oh you're right! Forgot that I could tweak this method too, thanks👌
